File: python-solutions/binarysearch/first_bad_version.py
def first_bad_version(n):
    """
    :type n: int
    :rtype: int
    """
    left = 1
    right = n

    while left <= right:
        mid = (left + right) / 2

        if isBadVersion(mid):
            right = mid - 1
        else:
            left = mid + 1

    return left


# [1, 2, 3, 4, 5], bad = 4
# left = 1 , right = 5, mid = 3
# left = 1, right = 2, mid = 1
# left = 2, right = 2, mid = 2
# left = 3, right = 2

def isBadVersion(version):
    return version > 3  # just for check


def main():
    first_bad_version(5)


if __name__ == "__main__":
    main()

File: java-solutions/src/main/java/com/lc/solutions/binarysearch/FirstBadVersion.java
package com.lc.solutions.binarysearch;

public class FirstBadVersion
{
    public static int firstBadVersion(int n)
    {
        int start = 1;
        int end = n - 1;

        while (start <= end)
        {
            int middle = (start + (end - start) / 2);

            if (isBadVersion(middle))
            {
                end = middle - 1;
            }
            else
            {
                start = middle + 1;
            }
        }
        return start;
    }

    private static boolean isBadVersion(int version)
    {
        return version > 3;
    }

    public static void main(String[] args)
    {
        System.out.println(firstBadVersion(5));
    }
}
